x/exchange/keeper/grpc_query.go

```diff
@@ -545,3 +545,132 @@ func (k QueryServer) ValidateManageFees(goCtx context.Context, req *exchange.Que
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	resp := &exchange.QueryGetPaymentResponse{}
+	resp.Payment, _ = k.Keeper.GetPayment(ctx, source, req.ExternalId)
+	if resp.Payment == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "no payment found with source %s and external id %q",
+			req.Source, req.ExternalId)
+	}
+
+	return resp, nil
+}
+
+// GetPaymentsWithSource gets all payments with a specific source account.
+func (k QueryServer) GetPaymentsWithSource(goCtx context.Context, req *exchange.QueryGetPaymentsWithSourceRequest) (*exchange.QueryGetPaymentsWithSourceResponse, error) {
+	if req == nil || len(req.Source) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	source, err := sdk.AccAddressFromBech32(req.Source)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid source %q: %v", req.Source, err)
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	keyPrefix := GetKeyPrefixPaymentsForSource(source)
+	preStore := prefix.NewStore(k.getStore(ctx), keyPrefix)
+
+	resp := &exchange.QueryGetPaymentsWithSourceResponse{}
+	var pageErr error
+	resp.Pagination, pageErr = query.Paginate(preStore, req.Pagination, func(_, value []byte) error {
+		// Only add it to the result if we can read it. This might result in fewer results than the limit,
+		// but at least one bad entry won't block others by causing the whole thing to return an error.
+		payment, _ := k.parsePaymentStoreValue(value)
+		if payment != nil {
+			resp.Payments = append(resp.Payments, payment)
+		}
+		return nil
+	})
+
+	if pageErr != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "error iterating payments with source %s: %v", req.Source, pageErr)
+	}
+
+	return resp, nil
+}
+
+// GetPaymentsWithTarget gets all payments with a specific target account.
+func (k QueryServer) GetPaymentsWithTarget(goCtx context.Context, req *exchange.QueryGetPaymentsWithTargetRequest) (*exchange.QueryGetPaymentsWithTargetResponse, error) {
+	if req == nil || len(req.Target) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "empty request")
+	}
+	target, err := sdk.AccAddressFromBech32(req.Target)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "invalid target %q: %v", req.Target, err)
+	}
+
+	ctx := sdk.UnwrapSDKContext(goCtx)
+	keyPrefix := GetIndexKeyPrefixTargetToPayments(target)
+	store := k.getStore(ctx)
+	preStore := prefix.NewStore(store, keyPrefix)
+
+	resp := &exchange.QueryGetPaymentsWithTargetResponse{}
+	var pageErr error
+	resp.Pagination, pageErr = query.Paginate(preStore, req.Pagination, func(keySuffix, _ []byte) error {
+		// Only add it to the result if we can read it. This might result in fewer results than the limit,
+		// but at least one bad entry won't block others by causing the whole thing to return an error.
+		if source, externalID, pErr := ParseIndexKeySuffixTargetToPayment(keySuffix); pErr == nil {
+			payment, _ := k.getPaymentFromStore(store, source, externalID)
```

Comment:
Same as above. It would be helpful to know something is wrong rather than just masking it. This still allows you to achieve your goal of an entry not blocking others, but it will at least provide us some info when it is not correctly behaving.